### PR TITLE
The new core

### PR DESCRIPTION
**Pull request by jadrubio**
### Description

-

### Related Jira Tickets

-

### Comments 

-

Please ensure all items are complete before requesting a review:

- [ ] **Code Completed**: All code for the feature or bugfix has been implemented.
- [ ] **Unit Tests Added**: Unit tests are written for all new functionality, covering edge cases where applicable.
- [ ] **Integration Tests (where necessary)**: Integration tests are added for interactions between modules or components.
- [ ] **All Tests Passing**: All unit and integration tests are passing.
- [ ] **Type Safety Checked**: Ensure TypeScript types are correctly defined and used. Avoid `any` types without proper justification.
- [ ] **Typedoc Comments Added**: Ensure all functions, classes, interfaces, and types are documented using TypeDoc syntax.
- [ ] **Code Formatting**: All code files are formatted according to the project’s standards (e.g., Prettier, ESLint) and no linting errors remain.
- [ ] **No Console Logs**: Remove any `console.log` or other debugging output from the code.
- [ ] **Documentation Updated**: Update relevant project documentation (e.g., README, configuration guides) if necessary.

---

### CODE REVIEW

File: src/__tests__/README.md [hunk 1]
```diff
@@ -0,0 +1,273 @@
+# Testing Philosophy & Practices for @just-in/core
```
Comment (numerator) [inline src/__tests__/README.md]: I love this! I'll need to spend more time with it to wrap my head around the parts about mocking and spying and/or if those parts might need a bit more, but overall the philosophy and approach is great and I love that you put in the effort to get this document off the ground!

File: src/logger/global.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+      console.error(line);
+      break;
+    case 'INFO':
+    default:
+      console.log(line);
+      break;
+  }
+}
+
+let _globalMinLevel: string = (process.env.LOG_LEVEL as string) ?? 'DEBUG';
```
Comment (peiyaoh) [inline src/logger/global.ts:40]: [Non blocking] Is it correct to understand it as the default min level that will be applied if nothing is specified? In other words, it is the default value that can be used globally across an application.

File: src/data-manager/mongo/mongo.helpers.ts [hunk 3]
```diff
@@ -1,28 +1,112 @@
 import * as mongoDB from 'mongodb';
-import { Log } from '../../logger/logger-manager'
+import { NO_ID } from '../data-manager.constants';
+import { createLogger } from '../../logger/logger';
 
+const Log = createLogger({
```
Comment (peiyaoh) [inline src/data-manager/mongo/mongo.helpers.ts:5]: It's very exciting/lovely to see we can declare logger with context to scope it. Very easy to use and understand.

File: src/data-manager/change-listener.manager.ts [hunk 4]
```diff
@@ -102,10 +105,7 @@ export class ChangeListenerManager extends EventEmitter {
    * @param {string} collectionName - The name of the collection.
    * @param {CollectionChangeType} changeType - The type of changes to stop listening for.
    */
-  public removeChangeListener(
-    collectionName: string,
-    changeType: CollectionChangeType
-  ): void {
+  public removeChangeListener(collectionName: string, changeType: CollectionChangeType): void {
```
Comment (peiyaoh) [inline src/data-manager/change-listener.manager.ts:108]: [For the body] It takes some time for me to understand why we need a custom cleanup for/before destruction, but now I can appreciate the thinking behind it.

File: src/data-manager/data-manager.ts [hunk 5]
```diff
@@ -46,39 +50,48 @@ class DataManager extends EventEmitter {
...
+    this.checkInitialization();
+    await this.db.ensureStore(storeName, options as any);
+  }
+
+  /**
+   * Ensures indexes exist on a store (idempotent by name and key).
+   * @param {string} storeName - The collection/table name.
+   * @param {Array<{name?: string; key: unknown; unique?: boolean; partialFilterExpression?: unknown}>} indexes
+   */
+  public async ensureIndexes(
```
Comment (peiyaoh) [inline src/data-manager/data-manager.ts:84]: Love the support of index! It will be super useful.

File: src/data-manager/data-manager.ts [hunk 6]
```diff
@@ -46,39 +50,48 @@ class DataManager extends EventEmitter {
...
    */
   public async init(dbType: DBType = DBType.MONGO): Promise<void> {
-    Log.dev('Entering DM.init, isInitialized:', this.isInitialized);
     try {
       if (this.getInitializationStatus() && dbType === DBType.MONGO) return;
       if (dbType !== DBType.MONGO) {
         throw new Error('MongoDB is the only supported DB type');
       }
       await this.db.init();
       this.isInitialized = true;
```
Comment (peiyaoh) [inline src/data-manager/data-manager.ts:63]: I was wondering whether we actually need isInitialized in data-manager, instead of letting the db (mongo-manager in this case) provide that status. It seems that the status is entirely dependent on the underlying DB's status. In that case, would it be better to have a single source of truth if no additional initialization is required?

File: src/data-manager/data-manager.ts [hunk 7]
```diff
@@ -120,26 +131,22 @@ class DataManager extends EventEmitter {
...
-    item: object
-  ): Promise<object | null> {
+  public async addItemToCollection(collectionName: string, item: object): Promise<object | null> {
     try {
       this.checkInitialization();
       const id = await this.db.addItemToCollection(collectionName, item);
       const newItem = { id, ...item };
 
       if (collectionName === USERS) {
         this.emit('userAdded', newItem);
```
Comment (peiyaoh) [inline src/data-manager/data-manager.ts:141]: Should we standardize it so that it could in theory work for any collections that need these events?

Something like 

```typescript
const eventCollection = [USERS];

 if (collectionName in eventCollection) {
        this.emit('${collectionName}-addItem', newItem);
}
```

File: src/data-manager/data-manager.ts [hunk 8]
```diff
@@ -289,24 +289,21 @@ class DataManager extends EventEmitter {
    */
   public async findItemsInCollection<T>(
     collectionName: string,
-    criteria: Record<string, any>
+    criteria: Record<string, any>,
```
Comment (peiyaoh) [inline src/data-manager/data-manager.ts:292]: Update the comment to reflect `Record<string, any>`

File: src/data-manager/data-manager.type.ts [hunk 9]
```diff
@@ -12,10 +12,7 @@ export enum CollectionChangeType {
...
 
 export type CollectionChangeListener = {
-  (document: {
-    fullDocument: object;
-    updateDescription?: object;
-  }): Promise<void>;
+  (document: { fullDocument: object; updateDescription?: object }): Promise<void>;
 };
 
 export type CollectionChangeNotifier = {
```
Comment (peiyaoh) [inline src/data-manager/data-manager.type.ts:18]: I cannot find how we used it. If that is the case, consider remove it?

File: src/user-manager/user-manager.ts [hunk 10]
```diff
@@ -59,8 +60,7 @@ const shutdown = () => {
 const refreshCache = async (): Promise<void> => {
   _checkInitialization();
   _users.clear();
-  const userDocs =
-    (await dm.getAllInCollection<JUser>(USERS)) || [];
+  const userDocs = (await dm.getAllInCollection<JUser>(USERS)) || [];
```
Comment (peiyaoh) [inline src/user-manager/user-manager.ts:63]: This is minor. It's a bit strange that we are getting what is supposed to be `JUser` and then still have to transform it into `JUser` below. Isn't the transformation already provided by the database manager (through mongo-manager)?

File: src/user-manager/user-manager.ts
Comment (peiyaoh) [inline src/user-manager/user-manager.ts:1]: I was wondering about your current thoughts on the input validation for different methods in this module. Should it be done differently? We discussed this before, and I contributed some, if not most, of them. Now that I looked at it a second time, it seems a lot. It would be great to hear your thoughts.
